src/vl-tooltip.js

```diff
@@ -0,0 +1,624 @@
+
+          // supplement the field with fieldDef, timeFormat and numberFormat
+          var supplementedField = supplementField(field, fieldDef, timeFormat, numberFormat);
+
+          supplementedFields.push(supplementedField);
+        })
+      }
+    }
+
+    // TODO(zening): supplement binned fields
+
+    options.fields = supplementedFields;
+
+    return options;
+  }
+
+  /**
+  * Find a field in fields that matches a fieldDef
+  * If the fieldDef is aggregated, find a field that matches the field name and
+  * the aggregation of the fieldDef.
+  * If the fieldDef is not aggregated, find a field that matches the field name.
+  * @return the matching field, or undefined if no match was found
+  */
+  function getField(fields, fieldDef) {
+    if (!fieldDef || !fields || fields.length <= 0) return;
+
+    var matchedField = undefined;
+
+    // if aggregate, match field name and aggregate operation
+    if (fieldDef.aggregate) {
+      // try find the perfect match: field name equals, aggregate operation equals
+      fields.forEach(function(field) {
+        if (!matchedField && field.field === fieldDef.field && field.aggregate === fieldDef.aggregate) {
+          matchedField = field;
+        }
+      });
+
+      // try find the second-best match: field name equals, field.aggregate = undefined
+      if (!matchedField) {
+        fields.forEach(function(field) {
+          if (!matchedField && field.field === fieldDef.field && field.aggregate === undefined) {
+            matchedField = field;
+          }
+        });
+      }
+    }
+    // if not aggregate, just match field name
+    else {
+      fields.forEach(function(field) {
+        if (!matchedField && field.field === fieldDef.field) {
+          matchedField = field;
+        }
+      });
+    }
+
+    return matchedField;
+  }
+
+  /**
+  * Find a fieldDef that matches field
+  * A matching fieldDef should have the same field name as field.
+  * If the matching fieldDef is aggregated, the aggregation should not contradict
+  * with that of the field.
+  * @return the matching fieldDef, or undefined if no match was found
+  */
+  function getFieldDef(fieldDefs, field) {
+    if (!field || !field.field || !fieldDefs) return;
+
+    var matchedFieldDef = undefined;
+
+    // field name should match, aggregation should not disagree
+    fieldDefs.forEach(function(fieldDef) {
+      if (!matchedFieldDef && fieldDef.field === field.field) {
+        if (fieldDef.aggregate) {
+          if (fieldDef.aggregate === field.aggregate || field.aggregate === undefined) {
+            matchedFieldDef = fieldDef;
+          }
+        }
+        else {
+          matchedFieldDef = fieldDef;
+        }
+      }
+    });
+
+    return matchedFieldDef;
+  }
+
+  /**
+  * Supplement a field (from options) with a fieldDef, timeFormat and numberFormat
+  * Either field or fieldDef can be undefined, but they cannot both be undefined.
```

Comment:
Here, `field` is one of `options.fields`. It's an object that may contain custom field title and format specified by the user. We are supplementing this `field` object with `fieldDef`, `timeFormat`, and `numberFormat`, which all come from vlSpec.